File: kora-app-annotation-processor/src/main/java/ru/tinkoff/kora/kora/app/annotation/processor/component/ComponentDependencyHelper.java
package ru.tinkoff.kora.kora.app.annotation.processor.component;

import com.palantir.javapoet.ParameterizedTypeName;
import com.palantir.javapoet.TypeName;
import org.jspecify.annotations.Nullable;
import ru.tinkoff.kora.annotation.processor.common.*;
import ru.tinkoff.kora.kora.app.annotation.processor.ProcessingContext;
import ru.tinkoff.kora.kora.app.annotation.processor.component.DependencyClaim.DependencyClaimType;
import ru.tinkoff.kora.kora.app.annotation.processor.declaration.ComponentDeclaration;

import javax.lang.model.element.Element;
import javax.lang.model.element.ElementKind;
import javax.lang.model.element.ExecutableElement;
import javax.lang.model.type.DeclaredType;
import javax.lang.model.type.ExecutableType;
import javax.lang.model.type.TypeMirror;
import java.util.ArrayList;
import java.util.List;

public class ComponentDependencyHelper {

    public static List<DependencyClaim> parseDependencyClaims(ProcessingContext ctx, ComponentDeclaration componentDeclaration) {
        // TODO switch
        if (componentDeclaration instanceof ComponentDeclaration.FromModuleComponent moduleComponent) {
            var element = moduleComponent.method();
            var result = new ArrayList<DependencyClaim>(moduleComponent.methodParameterTypes().size() + 1);
            for (int i = 0; i < moduleComponent.methodParameterTypes().size(); i++) {
                var parameterType = moduleComponent.methodParameterTypes().get(i);
                var parameterElement = element.getParameters().get(i);
                var tags = TagUtils.parseTagValue(parameterElement);
                var isNullable = CommonUtils.isNullable(parameterElement);
                result.add(parseClaim(element, parameterType, tags, isNullable));
            }
            return result;
        } else if (componentDeclaration instanceof ComponentDeclaration.DiscoveredAsDependencyComponent discoveredAsDependency) {
            var element = discoveredAsDependency.constructor();
            var type = (ExecutableType) ctx.types.asMemberOf(discoveredAsDependency.type(), element);
            var result = new ArrayList<DependencyClaim>(element.getParameters().size());
            for (int i = 0; i < type.getParameterTypes().size(); i++) {
                var parameterType = type.getParameterTypes().get(i);
                var parameterElement = element.getParameters().get(i);
                var tags = TagUtils.parseTagValue(parameterElement);
                var isNullable = CommonUtils.isNullable(parameterElement);
                result.add(parseClaim(element, parameterType, tags, isNullable));
            }
            return result;
        } else if (componentDeclaration instanceof ComponentDeclaration.AnnotatedComponent annotated) {
            var element = annotated.constructor();
            var type = (ExecutableType) annotated.constructor().asType();
            var result = new ArrayList<DependencyClaim>(element.getParameters().size());
            for (int i = 0; i < annotated.methodParameterTypes().size(); i++) {
                var parameterType = annotated.methodParameterTypes().get(i);
                var parameterElement = element.getParameters().get(i);
                var tags = TagUtils.parseTagValue(parameterElement);
                var isNullable = CommonUtils.isNullable(parameterElement);
                result.add(parseClaim(element, parameterType, tags, isNullable));
            }
            return result;
        } else if (componentDeclaration instanceof ComponentDeclaration.FromExtensionComponent fromExtension) {
            var result = new ArrayList<DependencyClaim>(fromExtension.dependencyTypes().size() + 1);
            var executable = fromExtension.source().getKind() == ElementKind.METHOD || fromExtension.source().getKind() == ElementKind.CONSTRUCTOR
                ? (ExecutableElement) fromExtension.source()
                : null;
            for (int i = 0; i < fromExtension.dependencyTypes().size(); i++) {
                var parameterType = fromExtension.dependencyTypes().get(i);
                var tags = fromExtension.dependencyTags().get(i);
                var element = executable == null ? null : executable.getParameters().get(i);
                var isNullable = element != null && CommonUtils.isNullable(element);
                result.add(parseClaim(fromExtension.source(), parameterType, tags, isNullable));
            }
            return result;
        }
        throw new IllegalArgumentException();
    }

    public static DependencyClaim parseClaim(Element sourceElement, TypeMirror parameterType, @Nullable String tag, boolean isNullable) {
        if (TypeParameterUtils.hasRawTypes(parameterType)) {
            throw new ProcessingErrorException("Components with raw types can break dependency resolution in unpredictable way so they are forbidden", sourceElement);
        }

        var typeName = TypeName.get(parameterType);
        if (typeName instanceof ParameterizedTypeName ptn && parameterType instanceof DeclaredType dt) {
            if (ptn.rawType().canonicalName().equals(CommonClassNames.typeRef.canonicalName())) {
                return new DependencyClaim(dt.getTypeArguments().getFirst(), tag, DependencyClaimType.TYPE_REF);
            }
            if (ptn.rawType().canonicalName().equals(CommonClassNames.all.canonicalName())) {
                if (ptn.typeArguments().getFirst() instanceof ParameterizedTypeName allOfType && dt.getTypeArguments().getFirst() instanceof DeclaredType allOfTypeName) {
                    if (allOfType.rawType().canonicalName().equals(CommonClassNames.valueOf.canonicalName())) {
                        return new DependencyClaim(allOfTypeName.getTypeArguments().getFirst(), tag, DependencyClaimType.ALL_OF_VALUE);
                    }
                    if (allOfType.rawType().canonicalName().equals(CommonClassNames.promiseOf.canonicalName())) {
                        return new DependencyClaim(allOfTypeName.getTypeArguments().getFirst(), tag, DependencyClaimType.ALL_OF_PROMISE);
                    }
                }
                return new DependencyClaim(dt.getTypeArguments().getFirst(), tag, DependencyClaimType.ALL_OF_ONE);
            }
            if (ptn.rawType().canonicalName().equals(CommonClassNames.valueOf.canonicalName())) {
                if (isNullable) {
                    return new DependencyClaim(dt.getTypeArguments().getFirst(), tag, DependencyClaimType.NULLABLE_VALUE_OF);
                } else {
                    return new DependencyClaim(dt.getTypeArguments().getFirst(), tag, DependencyClaimType.VALUE_OF);
                }
            }
            if (ptn.rawType().canonicalName().equals(CommonClassNames.promiseOf.canonicalName())) {
                if (isNullable) {
                    return new DependencyClaim(dt.getTypeArguments().getFirst(), tag, DependencyClaimType.NULLABLE_PROMISE_OF);
                } else {
                    return new DependencyClaim(dt.getTypeArguments().getFirst(), tag, DependencyClaimType.PROMISE_OF);
                }
            }
        }
        if (isNullable) {
            return new DependencyClaim(parameterType, tag, DependencyClaimType.ONE_NULLABLE);
        } else {
            return new DependencyClaim(parameterType, tag, DependencyClaimType.ONE_REQUIRED);
        }
    }

}

File: kora-app-symbol-processor/src/main/kotlin/ru/tinkoff/kora/kora/app/ksp/component/ComponentDependencyHelper.kt
package ru.tinkoff.kora.kora.app.ksp.component

import com.google.devtools.ksp.symbol.KSAnnotated
import com.google.devtools.ksp.symbol.KSType
import com.squareup.kotlinpoet.ParameterizedTypeName
import com.squareup.kotlinpoet.ksp.toTypeName
import ru.tinkoff.kora.kora.app.ksp.declaration.ComponentDeclaration
import ru.tinkoff.kora.ksp.common.AnnotationUtils.isAnnotationPresent
import ru.tinkoff.kora.ksp.common.CommonClassNames
import ru.tinkoff.kora.ksp.common.TagUtils
import ru.tinkoff.kora.ksp.common.exception.ProcessingError
import ru.tinkoff.kora.ksp.common.exception.ProcessingErrorException
import javax.tools.Diagnostic

object ComponentDependencyHelper {
    fun parseDependencyClaim(declaration: ComponentDeclaration): List<DependencyClaim> {
        when (declaration) {
            is ComponentDeclaration.FromModuleComponent -> {
                val element = declaration.method
                val result = ArrayList<DependencyClaim>(declaration.methodParameterTypes.size + 1)
                for (i in 0 until declaration.methodParameterTypes.size) {
                    val parameterType = declaration.methodParameterTypes[i]
                    val parameterElement = element.parameters[i]
                    val tags = TagUtils.parseTagValue(parameterElement)
                    result.add(parseClaim(parameterType, tags, declaration.method.parameters[i]))
                }
                return result
            }

            is ComponentDeclaration.AnnotatedComponent -> {
                val element = declaration.constructor
                val result = ArrayList<DependencyClaim>(declaration.methodParameterTypes.size)
                for (i in 0 until declaration.methodParameterTypes.size) {
                    val parameterType = declaration.methodParameterTypes[i]
                    val tags = TagUtils.parseTagValue(element.parameters[i])
                    result.add(parseClaim(parameterType, tags, declaration.constructor.parameters[i]))
                }
                return result
            }

            is ComponentDeclaration.DiscoveredAsDependencyComponent -> {
                val element = declaration.constructor
                val constructorParameterTypes = if (declaration.classDeclaration.typeParameters.isEmpty()) {
                    element.parameters.asSequence().map { it.type.resolve() }
                } else {
                    // this will fail but there's nothing we can do about it
                    element.asMemberOf(declaration.type).parameterTypes.asSequence().mapNotNull { it!! }
                }
                val result = ArrayList<DependencyClaim>(element.parameters.size)
                for ((parameter, parameterType) in element.parameters.asSequence().zip(constructorParameterTypes)) {
                    val tags = TagUtils.parseTagValue(parameter)
                    result.add(parseClaim(parameterType, tags, element))
                }
                return result
            }

            is ComponentDeclaration.FromExtensionComponent -> {
                val result = ArrayList<DependencyClaim>(declaration.methodParameterTypes.size)
                for ((type, tags) in declaration.methodParameterTypes.zip(declaration.methodParameterTags)) {
                    result.add(parseClaim(type, tags, declaration.source))
                }
                return result
            }

            else -> throw IllegalArgumentException()
        }
    }


    fun parseClaim(parameterType: KSType, tag: String?, element: KSAnnotated): DependencyClaim {
        if (parameterType.isError) {
            throw ProcessingErrorException(ProcessingError("Dependency type parameter is not resolvable in the current round of processing: $element\nTry disabling Kora KSP processor dependency and compile without it to check for errors in your codebase (Kotlin and KSP compiler work only this way)", element, Diagnostic.Kind.WARNING))
        }
        val typeName = try {
            parameterType.toTypeName()
        } catch (e: IllegalArgumentException) {
            throw ProcessingErrorException(ProcessingError("Dependency type parameter is not resolvable in the current round of processing: $element\nTry disabling Kora KSP processor dependency and compile without it to check for errors in your codebase (Kotlin and KSP compiler work only this way)", element, Diagnostic.Kind.WARNING))
        }
        if (typeName is ParameterizedTypeName) {
            val firstTypeParam = parameterType.arguments[0].type!!.resolve()
            if (typeName.rawType == CommonClassNames.typeRef) {
                return DependencyClaim(firstTypeParam, tag, DependencyClaim.DependencyClaimType.TYPE_REF)
            }
            if (typeName.rawType == CommonClassNames.all) {
                val allOf = typeName.typeArguments[0]
                if (allOf is ParameterizedTypeName) {
                    if (allOf.rawType == CommonClassNames.valueOf) {
                        return DependencyClaim(firstTypeParam.arguments[0].type!!.resolve(), tag, DependencyClaim.DependencyClaimType.ALL_OF_VALUE)
                    }
                    if (allOf.rawType == CommonClassNames.promiseOf) {
                        return DependencyClaim(firstTypeParam.arguments[0].type!!.resolve(), tag, DependencyClaim.DependencyClaimType.ALL_OF_PROMISE)
                    }
                }
                return DependencyClaim(firstTypeParam, tag, DependencyClaim.DependencyClaimType.ALL)
            }
            if (typeName.rawType == CommonClassNames.valueOf) {
                if (parameterType.isMarkedNullable || element.isAnnotationPresent(CommonClassNames.nullable)) {
                    return DependencyClaim(firstTypeParam, tag, DependencyClaim.DependencyClaimType.NULLABLE_VALUE_OF)
                } else {
                    return DependencyClaim(firstTypeParam, tag, DependencyClaim.DependencyClaimType.VALUE_OF)
                }
            }
            if (typeName.rawType == CommonClassNames.promiseOf) {
                if (parameterType.isMarkedNullable || element.isAnnotationPresent(CommonClassNames.nullable)) {
                    return DependencyClaim(firstTypeParam, tag, DependencyClaim.DependencyClaimType.NULLABLE_PROMISE_OF)
                } else {
                    return DependencyClaim(firstTypeParam, tag, DependencyClaim.DependencyClaimType.PROMISE_OF)
                }
            }
        }
        if (parameterType.isMarkedNullable || element.isAnnotationPresent(CommonClassNames.nullable)) {
            return DependencyClaim(parameterType, tag, DependencyClaim.DependencyClaimType.NULLABLE_ONE)
        } else {
            return DependencyClaim(parameterType, tag, DependencyClaim.DependencyClaimType.ONE_REQUIRED)
        }
    }
}
